engine/api/lastupdate.go

```diff
@@ -0,0 +1,182 @@
+				return
+			}
+		case <-tick:
+			msg, err := cache.GetMessageFromSubscription(pubSub, c)
+			if err != nil {
+				log.Warning("lastUpdate.CacheSubscribe> Cannot get message %s: %s", msg, err)
+				time.Sleep(5 * time.Second)
+				continue
+			}
+			cacheMsgChan <- msg
+		}
+	}
+}
+
+// Start the broker
+func (b *LastUpdateBroker) Start(c context.Context, db gorp.SqlExecutor) {
+
+	for {
+		select {
+		case <-c.Done():
+			// Close all channels
+			for c, v := range b.clients {
+				delete(b.clients, c)
+				close(v.Queue)
+			}
+			if c.Err() != nil {
+				log.Warning("lastUpdate.CacheSubscribe> Exiting: %v", c.Err())
+				return
+			}
+		case s := <-b.newClients:
+			// Register new client
+			b.clients[s.UIID] = s
+		case msg := <-b.messages:
+			var lastModif sdk.LastModification
+			if err := json.Unmarshal([]byte(msg), &lastModif); err != nil {
+				log.Warning("lastUpdate.CacheSubscribe> Cannot unmarshal message: %s", msg)
+				continue
+			}
+
+			//Receive new message
+			for _, i := range b.clients {
+				if err := loadUserPermissions(db, i.User); err != nil {
+					log.Warning("lastUpdate.CacheSubscribe> Cannot load auser permission: %s", err)
+					continue
+				}
+
+				hasPermission := false
+				if i.User.Admin {
+					hasPermission = true
+				} else {
+				groups:
+					for _, g := range i.User.Groups {
+						hasPermission = false
+						switch lastModif.Type {
+						case sdk.ApplicationLastModificationType:
+							for _, ag := range g.ApplicationGroups {
+								if ag.Application.Name == lastModif.Name && ag.Application.ProjectKey == lastModif.Key {
+									hasPermission = true
+									break groups
+								}
+							}
+						case sdk.PipelineLastModificationType:
+							for _, pg := range g.PipelineGroups {
+								if pg.Pipeline.Name == lastModif.Name && pg.Pipeline.ProjectKey == lastModif.Key {
+									hasPermission = true
+									break groups
+								}
+							}
+						}
+					}
+				}
+
+				if hasPermission {
+					i.Queue <- msg
+				}
+			}
+		}
+	}
+
+}
+
+func (b *LastUpdateBroker) ServeHTTP(w http.ResponseWriter, r *http.Request, db *gorp.DbMap, c *businesscontext.Ctx) error {
+	// Make sure that the writer supports flushing.
+	f, ok := w.(http.Flusher)
+	if !ok {
+		http.Error(w, "Streaming unsupported!", http.StatusInternalServerError)
+		return nil
+	}
+
+	uuid, errS := sessionstore.NewSessionKey()
+	if errS != nil {
+		return sdk.WrapError(errS, "LastUpdateBroker.Serve> Cannot generate UUID")
+	}
+	messageChan := &LastUpdateBrokerSubscribe{
+		UIID:  string(uuid),
+		User:  c.User,
+		Queue: make(chan string),
+	}
+
+	// Add this client to the map of those that should receive updates
+	b.newClients <- messageChan
+
+	// Set the headers related to event streaming.
+	w.Header().Set("Content-Type", "text/event-stream")
+	w.Header().Set("Cache-Control", "no-cache")
+	w.Header().Set("Connection", "keep-alive")
+
+leave:
+	for {
+		select {
+		case <-w.(http.CloseNotifier).CloseNotify():
```

Comment:
it's a golang question: this can't panic ? what append if the cast does't match ?